src/leap/common/http.py

```diff
@@ -36,21 +36,141 @@
 from twisted.web.client import readBody
 from twisted.web.http_headers import Headers
 from twisted.web.iweb import IBodyProducer
+from twisted.web._newclient import HTTP11ClientProtocol as \
+    _HTTP11ClientProtocol
 
 
-def createPool(maxPersistentPerHost=10, persistent=True):
-    pool = HTTPConnectionPool(reactor, persistent)
+_defaultTimeout = 30
+
+
+class HTTP11ClientProtocol(_HTTP11ClientProtocol):
+    """
+    A timeout-able HTTP 1.1 client protocol, that is instantiated by the
+    HTTP11ClientFactory below.
+    """
+
+    def __init__(self, quiescentCallback, timeout):
+        """
+        Initialize the protocol.
+
+        :param quiescentCallback:
+        :type quiescentCallback: callable
+        :param timeout: A timeout requests made by this protocol.
+        :type timeout: float
+        """
+        _HTTP11ClientProtocol(quiescentCallback)
+        self._timeout = timeout
+
+    def request(self, request):
+        """
+        Issue request over self.transport and return a Deferred which
+        will fire with a Response instance or an error.
+
+        :param request: The object defining the parameters of the request to
+                        issue.
+        :type request: twisted.web._newclient.Request
+
+        :return: A deferred which fires after the request has finished.
+        :rtype: Deferred
+        """
+        d = _HTTP11ClientProtocol.request(self, request)
+        if self._timeout:
+            self._last_buffer_len = 0
+            timeoutCall = reactor.callLater(
+                self._timeout, self._doTimeout, request)
+            self._finishedRequest.addBoth(self._cancelTimeout, timeoutCall)
+        return d
+
+    def _doTimeout(self, request):
+        """
+        Give up the request because of a timeout.
+
+        :param request: The object defining the parameters of the request to
+                        issue.
+        :type request: twisted.web._newclient.Request
+        """
+        buffer_len = len(self._parser._buffer)
+        if not self._parser._everReceivedData \
+                or self._last_buffer_len == buffer_len:
+            self._giveUp(
+                defer.TimeoutError(
+                    "Getting %s took longer than %s seconds."
+                    % (request.absoluteURI, self._timeout)))
+        else:
+            self._last_buffer_len = buffer_len
+
+    def _cancelTimeout(self, passthrough, timeoutCall):
+        """
+        Cancel the request timeout, when it's finished.
+
+        :param passthrough: The result of previous callback.
+        :type passthrough: any
+        :param timeoutCall:
+        :type timeoutCall:
+        """
+        if timeoutCall.active():
+            timeoutCall.cancel()
+        return passthrough
+
+
+class HTTP11ClientFactory(_HTTP11ClientFactory):
+    """
+    A timeout-able HTTP 1.1 client protocol factory.
+    """
+
+    def __init__(self, quiescentCallback, timeout):
```

Comment:
I had decided not to, because we're the only ones using it, and should only be used internally. Don't see a point in making it more generic for now. But if you think it'd make it better for some use, I can add it.